### ISSUE TITLE
Bug for paths regarding YAML and HTML output files

### ISSUE THREAD

**Issue by riedel-ferringer** [OSVVM/OSVVM-Scripts#37]
In OsvvmRequiredSettings.tcl, you define three variables
```
variable VhdlReportsDirectory     ""
variable OsvvmYamlResultsFile     "OsvvmRun.yml"
variable TranscriptYamlFile       "OSVVM_transcript.yml"
```

I think this is not entirely correct, because the two yml files are missing the Directory-Prefix. I think this might be correct:
```
variable VhdlReportsDirectory     ""
variable OsvvmYamlResultsFile     [file join $VhdlReportsDirectory "OsvvmRun.yml"]
variable TranscriptYamlFile       [file join $VhdlReportsDirectory "OSVVM_transcript.yml"]
```
As VhdlReportsDirectory is set to "", this issue doesn't really matter. However, once you change VhdlReportsDirectory to something else, yml and html output doesn't work any more, because those files are not found.

**Comment by riedel-ferringer** [comment]
Having said that, one could think about moving VhdlReportsDirectory to the user-modifiable scritps :-)

**Comment by JimLewis** [comment]
Quick Answer:   
`file join` cannot be done there because CurrentSimulationDirectory is not determined at this point in time.   You can enter the simulator, cd, and then start a build.   In that case, the build is referenced to the directory in which the build was started.

**Comment by JimLewis** [comment]
Additional Musings about this issue:
The "Required Settings" is for items that if you change, bad things happen.   
VhdlReportsDirectory is used by both the scripts and VHDL code for temporary files.
Changing it would require changing both the scripts and the VHDL code.

Currently the files created in VhdlReportsDirectory are strictly temporary and only stay there if there is a crash (tool or OSVVM script) that circumvents the scripts from processing the files.   In those cases, having them in CurrentSimulationDirectory makes them easy to find.

Currently the only valid value for VhdlReportsDirectory is "".   This could change in the future, but we would need a use case that shows having temporary files created somewhere else is a good thing.

It is likely that simulator transcripts (log files) will start as OsvvmRun.log in the VhdlReportsDirectory / CurrentSimulationDirectory.   This would simplify addressing issues with log file creation in some simulators - such as running vsim -batch on windows/MSYS2:   
`vsim -batch |& tee OsvvmRun.log`

**Comment by JimLewis** [comment]
Is there something you are trying to accomplish here that I am missing?   If yes, please reopen the issue - I am closing it for now.

**Comment by riedel-ferringer** [comment]
> Additional Musings about this issue: The "Required Settings" is for items that if you change, bad things happen. 
> VhdlReportsDirectory is used by both the scripts and VHDL code for temporary files. Changing it would require changing both the scripts and the VHDL code.

I know :-) 
 
> Currently the files created in VhdlReportsDirectory are strictly temporary and only stay there if there is a crash (tool or OSVVM script) that circumvents the scripts from processing the files. In those cases, having them in CurrentSimulationDirectory makes them easy to find.

That's exaclty what I want to avoid. I thought it to be a better solution to place all files, temporary or not, in the same "subfolder" so that my main simdir is not affected. This is not a big deal - I was just playing around. I also changed the VHDL code so that the yaml files are generated accordingly.

> Currently the only valid value for VhdlReportsDirectory is "". This could change in the future, but we would need a use case that shows having temporary files created somewhere else is a good thing.

I understand that. But I found that since you do have the defined variable, `OsvvmYamlResultsFile` and `TranscriptYamlFile` should be located inside it. I'd even suggest to add a corresponding constant to the VHDL code so that a user could, potentially, change both the tcl and vhdl files to place the output whereever he wants. As I said, it's not a big deal - but I think those [file join] statements are missing anyway :-)

> It is likely that simulator transcripts (log files) will start as OsvvmRun.log in the VhdlReportsDirectory / CurrentSimulationDirectory. This would simplify addressing issues with log file creation in some simulators - such as running vsim -batch on windows/MSYS2: `vsim -batch |& tee OsvvmRun.log`

**Comment by JimLewis** [comment]
What else do you put in your simdir?

I treat simdir as a temporary.   The issue I have is that simulators put temporary files in there.   Hence, before starting a regression, I delete the entire directory and recreate it.   

That said, it would not be too much trouble to do what you requested.   In addition, before setting VhdlReportsDirectory  in the required settings, I can check to see if it is set or not - and only set it if it is not already set.   That way you can set it in your local defaults.   However that said, OSVVM will not recommend that others change it.

**Comment by riedel-ferringer** [comment]
I was unclear with the term 'simdir': I just mean the folder with the test harness, the test_ctrl entity, the testcases and the .pro files. And i would like to keep it clean. 

But don't get me wrong here - this is not a feature request. I just found that as `VhdlReportsDirectory` exists, it should be part of those other two variables as well. It seemed to me like a "minor bug" for the improbable and not-intended case where someone changes the default value of `VhdlReportsDirectory` (or in case you will decide to make it configurable in the future)

**Comment by JimLewis** [comment]
Currently OSVVM testbenches (not in any normal OSVVM library), there is an OsvvmTestCommonPkg.vhd that establishes the path to OsvvmLibraries - using the constant OSVVM_PATH_TO_TESTS.   In there the constant OSVVM_RESULTS_DIR is also defined.

**Comment by riedel-ferringer** [comment]
I'm not sure if I get all the implications here (well, I am sure that I don't :-) ), but I'll try to comment anyway. I'll try to summarize my point of view, sorry for being repetitive.

- I understand that OSVVM creates some files in the "current working/simulation" directory and copies the files away once finished. I'd prefer that files are generated where they are located in the end. For various reasons:
  -  If something goes wrong, the files are not moved and thus must be removed manually
  -  If something goes wrong, I cannot look for the files in their usual location and check their contents for errors because they are not there - they still are at the old location, which I have to know...
  - If there are scripts involved that e.g. post-process those files, either after or while they are generated, there are obvious problems. This seems constructed, but if you don't manually review HTML reports, grep-ing logfiles and generating reports etc. is a valid way to go.
- I know that there is a missing link between the tcl scripts and VHDL. VHDL must be informed about some paths by the tcl scripts, otherwise you'd have some constants/variables that must be defined twice, once for .tcl and once for .vhd
  - One could force a shared variable to the respective value. This probably doesn't work in all simulators. I have done this a while with ModelSim.
  - One would have a setting that must be defined twice. Well, it's not desireable but it's OK. if there is a hint/comment in the tcl/vhd file which states that the other location must be updated as well.

So long story short, in my opinion it would be most portable and usable if there is a vhd package which contains some possibly duplicated settings. The "default user" doesn't have to change anyway, the sophisticated user can configure it to his liking.

**Comment by JimLewis** [comment]
@riedel-ferringer   Deja.   :)

In the dev release, I have created a new package, OsvvmScriptSettingsPkg.vhd.    Right now the package declarative part is auto-generated when running OsvvmLibraries/osvvm/osvvm.pro.  It defines:

```vhdl
  constant OSVVM_HOME_DIRECTORY        : string := "C:/SynthWorks/Dev/_osvvm/OsvvmLibraries" ;
  constant OSVVM_OUTPUT_DIRECTORY      : string := "" ;
  constant OSVVM_BUILD_YAML_FILE       : string := "OsvvmRun.yml" ;
  constant OSVVM_TRANSCRIPT_YAML_FILE  : string := "OSVVM_transcript.yml" ;
```
Names are not set in stone yet.   The osvvm packages have been updated to use these settings.   

It is likely that the approach will evolve some before release - specifically I think these will work better as deferred constants.   I think that will help with the compilation rules a little.  I want to avoid triggering recompile of the whole OSVVM library.   There are already safeguards that keep the package from being updated if the settings do not change.   

The OSVVM_OUTPUT_DIRECTORY is set to the same value as $::osvvm::OsvvmTemporaryOutputDirectory.   This value is set in OsvvmDefaultSettings.tcl and LocalScriptDefaults.tcl.   I expect the default setting may be setting it to match $::osvvm::OutputBaseDirectory - for me, this is not the setting that I need.   Typically I run with OutputBaseDirectory as "", except, when running GHDL where I set it to "OSVVM".   Hence, I need OsvvmTemporaryOutputDirectory to be "" as it needs to be set to the same value for all simulators run out of the same directory (otherwise recompilation will be a problem).   

The ultimate goal of course is to always have the generated files in the correct locations - I suspect this got much better after the error handling was added - this allowed the report tools to at least move the files and try to run even after a failure.   The only time I see things not getting created properly is when the simulator crashes and leaves files open - if the old files are not closed properly, new ones cannot overwrite them.   I am working to address this.   When it does not get moved, I consider it an escape and it needs to be recorded as an error so it can be fixed.
